virtual_ship/instruments/argo.py

```diff
@@ -0,0 +1,168 @@
+    vertical_speed = Variable("vertical_speed", dtype=np.float32)
+    cycle_days = Variable("cycle_days", dtype=np.int32)
+    drift_days = Variable("drift_days", dtype=np.int32)
+
+
+def _argo_vertical_movement(particle, fieldset, time):
+    if particle.cycle_phase == 0:
+        # Phase 0: Sinking with vertical_speed until depth is drift_depth
+        particle_ddepth += (  # noqa See comment above about particle_* variables.
+            particle.vertical_speed * particle.dt
+        )
+        if particle.depth + particle_ddepth <= particle.drift_depth:
+            particle_ddepth = particle.drift_depth - particle.depth
+            particle.cycle_phase = 1
+
+    elif particle.cycle_phase == 1:
+        # Phase 1: Drifting at depth for drifttime seconds
+        particle.drift_age += particle.dt
+        if particle.drift_age >= particle.drift_days * 86400:
+            particle.drift_age = 0  # reset drift_age for next cycle
+            particle.cycle_phase = 2
+
+    elif particle.cycle_phase == 2:
+        # Phase 2: Sinking further to max_depth
+        particle_ddepth += particle.vertical_speed * particle.dt
+        if particle.depth + particle_ddepth <= particle.max_depth:
+            particle_ddepth = particle.max_depth - particle.depth
+            particle.cycle_phase = 3
+
+    elif particle.cycle_phase == 3:
+        # Phase 3: Rising with vertical_speed until at surface
+        particle_ddepth -= particle.vertical_speed * particle.dt
+        particle.cycle_age += (
+            particle.dt
+        )  # solve issue of not updating cycle_age during ascent
+        if particle.depth + particle_ddepth >= particle.min_depth:
+            particle_ddepth = particle.min_depth - particle.depth
+            particle.temperature = (
+                math.nan
+            )  # reset temperature to NaN at end of sampling cycle
+            particle.salinity = math.nan  # idem
+            particle.cycle_phase = 4
+        else:
+            particle.temperature = fieldset.T[
+                time, particle.depth, particle.lat, particle.lon
+            ]
+            particle.salinity = fieldset.S[
+                time, particle.depth, particle.lat, particle.lon
+            ]
+
+    elif particle.cycle_phase == 4:
+        # Phase 4: Transmitting at surface until cycletime is reached
+        if particle.cycle_age > particle.cycle_days * 86400:
+            particle.cycle_phase = 0
+            particle.cycle_age = 0
+
+    if particle.state == StatusCode.Evaluate:
+        particle.cycle_age += particle.dt  # update cycle_age
+
+
+def _keep_at_surface(particle, fieldset, time):
+    # Prevent error when float reaches surface
+    if particle.state == StatusCode.ErrorThroughSurface:
+        particle.depth = particle.min_depth
+        particle.state = StatusCode.Success
+
+
+def _check_error(particle, fieldset, time):
+    if particle.state >= 50:  # This captures all Errors
+        particle.delete()
```

Comment:
This is a copy from the original code for now. I propose we improve the API after the first cleanup round. Preferrably the function would fail when this happens I assume.